library/src/main/java/dev/udell/open/util/LogReader.kt

```diff
@@ -0,0 +1,248 @@
+    copies or substantial portions of the Software.
+    
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+import android.annotation.TargetApi
+import android.content.ActivityNotFoundException
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.os.Build
+import android.os.Looper
+import android.text.format.DateFormat
+import android.view.Gravity
+import android.widget.Toast
+import androidx.core.content.FileProvider
+import androidx.startup.Initializer
+import dev.udell.open.R
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.io.*
+
+/**
+ * Utility class to grab a copy of the current app's log(cat) and share it.
+ */
+@Suppress("unused")
+class LogReader(context: Context) {
+    private val appContext = context.applicationContext
+    private val fileOps = FileUtils()
+
+    /**
+     * Return the current log as one (big) string.
+     *
+     * @return the log, or `null` on failure
+     */
+    suspend fun getLog(): String? {
+        val log = getRawLog()
+        var baos: ByteArrayOutputStream? = null
+        withContext(Dispatchers.Default) {
+            runCatching {
+                baos = ByteArrayOutputStream(log?.available() ?: 0)
+                baos?.use { out ->
+                    log?.copyTo(out)
+                }
+            }.onFailure {
+                baos = null
+            }
+        }
+        return baos?.toString()
+    }
+
+    /**
+     * Make the log available to the system's sharing providers. With no parameters, it'll share the
+     * current log as plain text to any app on the system registered to receive such. Supplying
+     * parameters will bias the share toward email.
+     *
+     * @param emailRecipient If supplied, configures the sharing intent for emailing to this address.
+     * @param emailHeaders Optional text to prepend to the email.
+     * @param shouldZip If true, will compress the log text into a .zip file
+     * @return `Boolean` for success
+     */
+    suspend fun shareLog(
+        emailRecipient: String? = null,
+        emailHeaders: String? = null,
+        shouldZip: Boolean = false
+    ): Boolean {
+        var useZip: Boolean = shouldZip
+
+        // Grab the log from the system and save it to our working dir
+        val targetName = LOG_SUBDIR + '/' + getLogFilePrefix(appContext) +
+                DateFormat.format("yyyyMMdd_hhmmss", System.currentTimeMillis()) + ".log"
```

Comment:
Anytime I see a custom date format I feel compelled to ask: why not ISO 8601?

Reply:
No good reason, probably just that this code predates my awareness of ISO 8601! I've opened an issue for this and will make the change at some point.